supervisor/core/supervisor.go

```diff
@@ -0,0 +1,214 @@
+	for id, inst := range sv.instancesById {
+		if inst.Cmd.Process.Pid == pid {
+			return id, inst
+		}
+	}
+	return 0, nil
+}
+
+// StartInstance starts a new Instance with the specified parameters.
+// On fail returns 0, error.
+func (sv *Supervisor) StartInstance(name string, env []string, restartable bool) (int, error) {
+	// When Supervisor is terminating, we will lock "termMutex"
+	// to prevent new instances from starting during Supervisor termination.
+	sv.termMutex.RLock()
+	defer sv.termMutex.RUnlock()
+
+	// Form the path to the instance and check is it exists.
+	if name == "" {
+		return 0, errors.New(`The instance name is empty.`)
+	}
+	instPath := path.Join(sv.cfg.InstancesDir, name+".lua")
+	if _, err := exec.LookPath(instPath); err != nil {
+		return 0, err
+	}
+
+	// Start an Instance.
+	cmd := exec.Command(instPath)
+	cmd.Env = append(os.Environ(), env...)
+	inst := NewInstance(name, cmd, env, restartable)
+	if err := inst.Start(); err != nil {
+		return 0, err
+	}
+
+	return sv.addInstance(inst), nil
+}
+
+// RestartAfterTermInstance should be used to restart an instance in case
+// of an unexpected termination of the instance.
+// First of all, the method checks if the instance is alive. In this case,
+// the method returns the current Instance ID with an error.
+// If the Instance has been terminated and Instance.Restartable is true,
+// then the Instance will be restarted.
+// On success returns the Instance ID (0 invalid).
+func (sv *Supervisor) RestartAfterTermInstance(pid int) (int, error) {
+	// Find an Instance.
+	id, inst := sv.getInstanceByPid(pid)
+	if inst == nil {
+		return 0, errors.New("Unknown Instance.")
+	}
+
+	// We don't want to restart the Instance at the same time
+	// as StopAllInstance is running.
+	sv.termMutex.RLock()
+	defer sv.termMutex.RUnlock()
+
+	// If the Instance is alive something went wrong.
+	if inst.IsAlive() {
+		return id, errors.New("Instance is alive.")
+	}
+
+	if !inst.Restartable {
+		return id, errors.New("The Restart Instance flag is false.")
+	}
+
+	if err := inst.Restart(); err != nil {
+		return id, err
+	}
+
+	return id, nil
+}
+
+// StopInstance terminate an Instance by ID.
+func (sv *Supervisor) StopInstance(id int, force bool) error {
+	// When Supervisor is terminating, we will lock "termMutex"
+	// to prevent an instance termination from several places
+	// (see StopAllInstance).
+	sv.termMutex.RLock()
+	defer sv.termMutex.RUnlock()
+
+	inst := sv.getInstance(id)
+	if inst == nil {
+		return errors.New("Unknown instance with id " + strconv.Itoa(id))
+	}
+
+	if err := inst.Stop(sv.cfg.TermTimeout, force); err != nil {
+		return err
+	}
+	sv.deleteInstance(id)
+
+	return nil
+}
+
+// StopAllInstance terminate all Instances managed by the Supervisor.
+func (sv *Supervisor) StopAllInstance() {
+	// Disable start / stop Instances.
+	sv.termMutex.Lock()
+
+	// Start termination of all Instances.
+	var wg sync.WaitGroup
+	// We don't want to delete instances from the map
```

Comment:
Say more hard: `We shouldn't delete instances`.

Reply:
Also, you can say here that this locks `deleteInstance` function until we iterate over the map.

Reply:
> Say more hard: `We shouldn't delete instances`.

Прям так жёсенько)

Reply:
> Also, you can say here that this locks `deleteInstance` function until we iterate over the map.

Hah, you like this part of code.
 A similar comment is already present below:
```
	// Here you need to Unlock the mutex, because while it is Lock,
	// no one goroutines created in the for loop will be done
	// (see the "deleteInstance" method).
```

Reply:
It's placed below, so I didn't find it on reading the related code above. If I see this comment within the mutex lock I would understand this better.

Reply:
This fact is also mentioned before `RLock ()`.